smart_open/s3.py

```diff
@@ -605,31 +695,68 @@ def read(self, size=-1):
         # HTTP connection and try again.  Usually, a single retry attempt is
         # enough to recover, but we try multiple times "just in case".
         #
-        for attempt, seconds in enumerate([1, 2, 4, 8, 16], 1):
-            try:
-                if size == -1:
-                    binary = self._body.read()
-                else:
-                    binary = self._body.read(size)
-            except (
-                ConnectionResetError,
-                botocore.exceptions.BotoCoreError,
-                urllib3.exceptions.HTTPError,
-            ) as err:
-                logger.warning(
-                    '%s: caught %r while reading %d bytes, sleeping %ds before retry',
-                    self,
-                    err,
-                    size,
-                    seconds,
-                )
-                time.sleep(seconds)
+        binary_collected = []
+        bytes_read = 0  # == sum(map(len, binary_collected))
```

Comment:
could use `binary_collected = bytearray()` to which you can efficiently `+=`, see also https://github.com/piskvorky/smart_open/pull/711

it also has a `len` so you can get rid of bytes_read